File: backend/runtime/internal/repository/long_term_memory_qdrant.go
package repository

import (
	"context"
	"fmt"

	"luna-ai/backend/runtime/internal/infrastructure"
	"luna-ai/backend/runtime/internal/logger"
)

// LongTermMemoryQdrantRepo 封装 Qdrant 中长期记忆的向量检索操作
// 做什么：提供记忆摘要的向量 Upsert、检索和删除操作
// 为什么这样做：Qdrant 作为语义检索引擎，通过向量相似度快速定位相关的历史记忆
type LongTermMemoryQdrantRepo struct {
	client *infrastructure.QdrantClient
}

// NewLongTermMemoryQdrantRepo 创建 LongTermMemoryQdrantRepo 实例
func NewLongTermMemoryQdrantRepo(client *infrastructure.QdrantClient) *LongTermMemoryQdrantRepo {
	return &LongTermMemoryQdrantRepo{client: client}
}

// EnsureCollection 确保长期记忆集合存在
// 向量维度：1536（默认与 OpenAI text-embedding-ada-002 对齐）
func (r *LongTermMemoryQdrantRepo) EnsureCollection(ctx context.Context, vectorSize int) error {
	return r.client.EnsureCollection(ctx, infrastructure.QdrantCollectionLongTermMemories, vectorSize)
}

// SaveWithVector 保存长期记忆向量
// 输入：
//   - ctx: 上下文
//   - memoryID: 记忆 ID（对应 PG 表的主键）
//   - sessionID: 会话 ID
//   - vector: 记忆摘要的 Embedding 向量
//   - status: 记忆状态（MemoryStatusActive / MemoryStatusDeleted）
//
// 输出：error
func (r *LongTermMemoryQdrantRepo) SaveWithVector(ctx context.Context, memoryID string, sessionID string, vector []float64, status MemoryStatus) error {
	if status == "" {
		status = MemoryStatusActive
	}
	point := infrastructure.UpsertPoint{
		ID:     memoryID,
		Vector: vector,
		Payload: map[string]interface{}{
			"session_id": sessionID,
			"status":     string(status),
		},
	}
	if err := r.client.Upsert(ctx, infrastructure.QdrantCollectionLongTermMemories, []infrastructure.UpsertPoint{point}); err != nil {
		return fmt.Errorf("保存长期记忆向量失败 [memory_id=%s]: %w", memoryID, err)
	}
	logger.Info(ctx, "长期记忆向量已保存", "memory_id", memoryID, "session_id", sessionID)
	return nil
}

// SearchByVector 根据向量检索长期记忆
// 输入：
//   - ctx: 上下文
//   - vector: 查询向量
//   - topK: 返回 Top-K 结果
//
// 输出：[]QdrantSearchResult, error
// 边界条件：仅返回 payload 中 status=MemoryStatusActive 的记录需要在业务层过滤
func (r *LongTermMemoryQdrantRepo) SearchByVector(ctx context.Context, vector []float64, topK int) ([]infrastructure.QdrantSearchResult, error) {
	results, err := r.client.Search(ctx, infrastructure.QdrantCollectionLongTermMemories, vector, topK)
	if err != nil {
		return nil, fmt.Errorf("检索长期记忆向量失败: %w", err)
	}

	// 过滤掉 status!=MemoryStatusActive 的结果
	activeStatus := string(MemoryStatusActive)
	activeResults := make([]infrastructure.QdrantSearchResult, 0, len(results))
	for _, result := range results {
		if status, ok := result.Payload["status"].(string); ok && status == activeStatus {
			activeResults = append(activeResults, result)
		}
	}

	logger.Info(ctx, "长期记忆向量检索完成", "hits", len(activeResults), "top_k", topK)
	return activeResults, nil
}

// SoftDeleteByMemoryID 根据记忆 ID 软删除向量（更新 payload 中的 status）
func (r *LongTermMemoryQdrantRepo) SoftDeleteByMemoryID(ctx context.Context, memoryID string) error {
	// Qdrant 不支持直接修改 payload 中单个字段，需重新 Upsert
	// 使用空向量 + MemoryStatusDeleted 状态覆盖
	point := infrastructure.UpsertPoint{
		ID: memoryID,
		// 使用零值向量覆盖：后续搜索时不会被匹配到（余弦相似度极低）
		Vector: make([]float64, 1536),
		Payload: map[string]interface{}{
			"status": string(MemoryStatusDeleted),
		},
	}
	if err := r.client.Upsert(ctx, infrastructure.QdrantCollectionLongTermMemories, []infrastructure.UpsertPoint{point}); err != nil {
		return fmt.Errorf("软删除长期记忆向量失败 [memory_id=%s]: %w", memoryID, err)
	}
	logger.Info(ctx, "长期记忆向量已软删除", "memory_id", memoryID)
	return nil
}

File: backend/ai-service/app/repository/long_term_memory_qdrant.py
"""
Luna AI Qdrant 长期记忆存储库

做什么：封装 Qdrant 中长期记忆的向量检索操作。
为什么这样做：Qdrant 作为语义检索引擎，通过向量相似度快速定位相关的历史记忆。
输入输出：
    - LongTermMemoryQdrantRepo: 长期记忆向量存储库类
边界条件：
    - 仅返回 payload 中 status=MemoryStatus.ACTIVE.value 的记录
异常行为：
    - 数据库操作失败时抛出异常
"""

from typing import List

from app.infrastructure.qdrant import (
    QDRANT_COLLECTION_LONG_TERM_MEMORIES,
    QdrantClientWrapper,
    QdrantSearchResult,
    UpsertPoint,
)
from app.logger import logger
from app.repository.models import MemoryStatus
from app.rag.chunker import MemoryChunk
from app.utils.snowflake import generate_id


class LongTermMemoryQdrantRepo:
    """封装 Qdrant 中长期记忆的向量检索操作"""

    def __init__(self, qdrant_client: QdrantClientWrapper):
        self.client = qdrant_client

    async def ensure_collection(self, vector_size: int) -> None:
        """
        确保长期记忆集合存在
        向量维度：768（默认与 BGE-base-zh-v1.5 对齐）
        """
        await self.client.ensure_collection(QDRANT_COLLECTION_LONG_TERM_MEMORIES, vector_size)

    async def save_chunks_with_vectors(
        self,
        memory_id: str,
        session_id: str,
        chunks: List[MemoryChunk],
        vectors: List[List[float]],
        status: str = ""
    ) -> None:
        """
        做什么：批量将拆分后的 Chunk 及其向量存入 Qdrant。
        为什么这样做：实现细粒度的 RAG 切片存储，以便后续检索时能够进行 search_groups 分组折叠去重。
        输入：memory_id（PG记录主键），session_id，MemoryChunk 列表及其对应的向量列表。
        异常行为：Qdrant 写入异常抛出。
        """
        if not status:
            status = MemoryStatus.ACTIVE.value
            
        if len(chunks) != len(vectors):
            raise ValueError(f"chunks 数量 ({len(chunks)}) 与 vectors 数量 ({len(vectors)}) 不匹配")

        points = []
        for chunk, vector in zip(chunks, vectors):
            point_id = generate_id()  # 每个 Chunk 分配独立的 Snowflake ID
            
            point = UpsertPoint(
                id=point_id,
                vector=vector,
                payload={
                    "memory_id": memory_id,
                    "session_id": session_id,
                    "chunk_type": chunk.chunk_type.value,
                    "content": chunk.content,
                    "status": status,
                }
            )
            points.append(point)
            
        await self.client.upsert(QDRANT_COLLECTION_LONG_TERM_MEMORIES, points)
        logger.info(f"[TraceID:N/A] 长期记忆切片向量已批量保存 memory_id={memory_id} chunks_count={len(chunks)}")

    async def search_groups_by_vector(self, query_vector: List[float], top_k: int) -> List[str]:
        """
        做什么：利用 Qdrant search_groups 获取去重后的 memory_id 列表。
        为什么这样做：解决同一条 PostgreSQL 记录的多个 Chunk 同时挤占 Top-K 检索名额的问题。
        边界条件：仅返回 payload 中 status=MemoryStatus.ACTIVE.value 的分组
        """
        results = await self.client.search_groups(
            collection_name=QDRANT_COLLECTION_LONG_TERM_MEMORIES,
            query_vector=query_vector,
            group_by="memory_id",
            limit=top_k,
            group_size=1
        )
        
        active_status = MemoryStatus.ACTIVE.value
        memory_ids = []
        for group in results:
            # group.id 就是 group_by 字段的值，即 memory_id
            # group.hits 包含组内的得分最高的 chunk
            if group.hits and group.hits[0].payload and group.hits[0].payload.get("status") == active_status:
                memory_ids.append(str(group.id))
                
        logger.info(f"[TraceID:N/A] 长期记忆分组向量检索完成 groups={len(results)} active_hits={len(memory_ids)} top_k={top_k}")
        return memory_ids

    async def search_by_vector(self, vector: List[float], top_k: int) -> List[QdrantSearchResult]:
        """
        根据向量检索长期记忆
        边界条件：仅返回 payload 中 status=MemoryStatus.ACTIVE.value 的记录
        """
        results = await self.client.search(QDRANT_COLLECTION_LONG_TERM_MEMORIES, vector, top_k)
        
        # 过滤掉 status!=MemoryStatus.ACTIVE.value 的结果
        active_status = MemoryStatus.ACTIVE.value
        active_results = [
            res for res in results
            if res.payload.get("status") == active_status
        ]
        
        logger.info(f"长期记忆向量检索完成 hits={len(active_results)} top_k={top_k}")
        return active_results

    # TODO: soft_delete_by_memory_id & delete_vector 以后需要改成 scroll 取出所有 point 然后修改 / 删除
    # 因为现在同一个 memory_id 会对应多个 Qdrant Points。但本次改动重点在 RAG Chunking 和 search_groups。
